### DISCUSSION TITLE
Registration w/ excludeCredentials

### DISCUSSION THREAD

**Discussion by daniel-mezo** [webauthn4j/webauthn4j#638] [General]
Hi @ynojima,

We are testing WebAuthn scenarios with the help of the `ClientPlatform` class in the library, and we found something interesting. If the server properly populates the `excludeCredentials` with pre-registered credential ids in the options during the registration ceremony, we get an `IllegalStateException` from the lib.
After some further digging, I believe this is traced to the following code snippet in `WebAuthnModelAuthenticator`:
https://github.com/webauthn4j/webauthn4j/blob/master/webauthn4j-test/src/main/java/com/webauthn4j/test/authenticator/webauthn/WebAuthnModelAuthenticator.java#L177

```java
 // For each descriptor of excludeCredentialDescriptorList:
List<PublicKeyCredentialDescriptor> descriptors = makeCredentialRequest.getExcludeCredentialDescriptorList();
if (descriptors == null) {
	descriptors = Collections.emptyList();
}
for (PublicKeyCredentialDescriptor descriptor : descriptors) {
	PublicKeyCredentialSource publicKeyCredentialSource = lookup(descriptor.getId());
	// If looking up descriptor.id in this authenticator returns non-null, and the returned item's RP ID
	// and type match rpEntity.id and excludeCredentialDescriptorList.type respectively,
	// then obtain user consent for creating a new credential.
	// The method of obtaining user consent MUST include a test of user presence.
	if (publicKeyCredentialSource != null) {
		if (publicKeyCredentialSource.getRpId().equals(rpEntity.getId()) &&
				publicKeyCredentialSource.getType().equals(descriptor.getType())) {
			boolean userConsent = true;
			// If the user
			// confirms consent to create a new credential
			if (userConsent) {
				throw new InvalidStateException("");
			}
			// does not consent to create a new credential
			else {
				throw new NotAllowedException("User consent is required");
			}
		}
	}
}
```

Notice, that if there is at least one credential which was already registered by this authenticator impl, - and the server signals that in the exclude credential list - there will be an exception thrown by the library.

I would expect that the success of `userConsent` (i.e. the value of the boolean variable) can somehow be signalled programatically to the library by it's caller, or the library would assume user consent by default, but either way the lib would proceed to continue with the ceremony as normal if user consent was established successfully, not to throw an `IllegalStateException`. This might be a missing implementation or a bug, what do you think?

**Comment by ynojima** [comment]
`webauthn4j-test` is in `experimental` state and some parts lacks implementation sorry.
Is there needs for WebAuthn Authenticator/ClientPlatform emulator?

**Reply by daniel-mezo** [reply]
Thank you for the response, I didn't know it was experimental to be honest.
We use these classes (ClientPlatform / WebAuthnModelAuthenticator and classes that extend it) to integration-/system-test our implementations, so I believe it would be a _nice to have_ if there was a fully functional emulator, it would make our life easier for sure, we could create more nuanced test scenarios.
We can also work around this current issue (we can create new instances of Authenticators that clears their credential cache), but was just surprised by it a bit.

**Reply by ynojima** [reply]
For system test, Google Chrome and Selenium-java added WebAuthn virtual authenticator support. It may help you implement E2E test.
https://github.com/SeleniumHQ/selenium/issues/7829

I stop to add feature to `webauthn4j-test` recently because I'm developing full-featured CTAP authenticator in Kotlin, but it is not finished and made public yet, sorry.
